src/core/ipe-setup.c

```diff
@@ -0,0 +1,122 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+
+#include "conf-files.h"
+#include "copy.h"
+#include "fd-util.h"
+#include "fileio.h"
+#include "ipe-setup.h"
+#include "path-util.h"
+
+#define IPE_SECFS_DIR "/sys/kernel/security/ipe"
+#define IPE_SECFS_NEW_POLICY IPE_SECFS_DIR "/new_policy"
+#define IPE_SECFS_POLICIES IPE_SECFS_DIR "/policies/"
+
+int ipe_setup(void) {
+#if ENABLE_IPE
+        _cleanup_strv_free_ char **policies = NULL;
+        int r;
+
+        /* Very quick smoke tests first: this is in the citical, sequential boot path, and in most cases it
+         * is unlikely this will be configured, so do the fastest existence checks first and immediately
+         * return if there's nothing to do. */
+
+        if (access(IPE_SECFS_DIR, F_OK) < 0) {
+                log_debug_errno(errno, "IPE support is disabled in the kernel, ignoring: %m");
+                return 0;
+        }
+
+        bool found = false;
+        STRV_FOREACH(p, CONF_PATHS_NULSTR("ipe"))
+                if (dir_is_empty(p, /* ignore_hidden_or_backup= */ true) == 0) {
```

Comment:
Hm? strv and nulstr?

BTW, why not just check strv_isempty(policies) below?

Reply:
see above and below comments

Reply:
Not resolved? You're still using STRV_FOREACH against _NULSTR?

Reply:
I don't understand the point, sorry

Reply:
`STRV_FOREACH(p, CONF_PATHS_NULSTR("ipe"))`

I think it's quite obvious really...

Reply:
Sorry but it's not - this works fine. If you want some different concoction, please say precisely what, and I'll test it again

Reply:
i am puzzled this is supposed to work too? @YHNdnzj is right?

a nulstr is a series of nul-terminated strings, concatenated in memory. An strv otoh is an array of individually nul terminated strings. you generate a nulstr here, and you iterate through it via strv. how could that possibly work?

Reply:
@bluca You should use `CONF_PATHS_STRV()` instead of `CONF_PATHS_NULSTR()` when you're calling `STRV_FOREACH`. The alternative is to use `NULSTR_FOREACH()` with `CONF_PATHS_NULSTR()`

Reply:
> how could that possibly work?

I have no clue, but it does

> You should use `CONF_PATHS_STRV()` instead of `CONF_PATHS_NULSTR()` when you're calling `STRV_FOREACH()`. The alternative is to use `NULSTR_FOREACH()` with `CONF_PATHS_NULSTR()`

Thanks, testing

Reply:
Computer says no:

```
../src/core/ipe-setup.c: In function ‘ipe_setup’:
../src/basic/nulstr-util.h:12:65: error: passing argument 1 of ‘strchr’ from incompatible pointer type [-Wincompatible-pointer-types]
   12 |         for (typeof(*(l)) *(i) = (l); (i) && *(i); (i) = strchr((i), 0)+1)
      |                                                                 ^~~
      |                                                                 |
      |                                                                 char **
../src/core/ipe-setup.c:49:9: note: in expansion of macro ‘NULSTR_FOREACH’
   49 |         NULSTR_FOREACH(policy, policies) {
      |         ^~~~~~~~~~~~~~
In file included from ../src/basic/log.h:7,
                 from ../src/basic/macro.h:384,
                 from ../src/basic/conf-files.h:6,
                 from ../src/core/ipe-setup.c:3:
/usr/include/string.h:246:34: note: expected ‘const char *’ but argument is of type ‘char **’
  246 | extern char *strchr (const char *__s, int __c)
      |                      ~~~~~~~~~~~~^~~
../src/basic/nulstr-util.h:12:56: error: assignment to ‘char **’ from incompatible pointer type ‘char *’ [-Wincompatible-pointer-types]
   12 |         for (typeof(*(l)) *(i) = (l); (i) && *(i); (i) = strchr((i), 0)+1)
      |                                                        ^
../src/core/ipe-setup.c:49:9: note: in expansion of macro ‘NULSTR_FOREACH’
   49 |         NULSTR_FOREACH(policy, policies) {
      |         ^~~~~~~~~~~~~~
```

I had a quick search and everything else (eg: `validate_signature_userspace()` seems to use STRV_FOREACH just like I am doing here, are you guys sure about this?

Reply:
where does validate_signature_userspace() do that?

and yes, i am damn sure i know how this works. just read the code!

Reply:
It's not about `policies`. It's about CONF_PATHS_*. You must use the proper iterator for nulstr and strv variants

Reply:
Please say that directly and clearly next time - change X to Y on line Z. I really, really, really don't care about this sort of stuff, if it compiles and runs that's good enough, and sometimes even without the second part